core/interop/src/lib.rs

```diff
@@ -40,67 +46,251 @@ impl<T: IntoIterator<Item = (JsString, NativeFunction)> + Clone> IntoJsModule fo
+/// let f = unsafe {
+///     let x = x.clone();
+///     move |a: i32| *x.borrow_mut() += a
+/// };
+/// let f = unsafe { f.into_js_function_unsafe(&mut context) };
+/// f.call(&JsValue::undefined(), &[JsValue::from(1)], &mut context).unwrap();
+/// f.call(&JsValue::undefined(), &[JsValue::from(4)], &mut context).unwrap();
+/// assert_eq!(*x.borrow(), 5);
+/// ```
+pub trait UnsafeIntoJsFunction<Args, Ret>: private::IntoJsFunctionSealed<Args, Ret> {
     /// Converts the type into a JS function.
     ///
     /// # Safety
     /// This function is unsafe to ensure the callee knows the risks of using this trait.
     /// The implementing type must not contain any garbage collected objects.
-    unsafe fn into_js_function(self, context: &mut Context) -> NativeFunction;
+    unsafe fn into_js_function_unsafe(self, context: &mut Context) -> NativeFunction;
 }
 
-unsafe impl<T: FnMut() + 'static> IntoJsFunctionUnsafe for T {
-    unsafe fn into_js_function(self, _context: &mut Context) -> NativeFunction {
-        let cell = RefCell::new(self);
-        unsafe {
-            NativeFunction::from_closure(move |_, _, _| {
-                cell.borrow_mut()();
-                Ok(JsValue::undefined())
-            })
+/// The safe equivalent of the [`UnsafeIntoJsFunction`] trait.
+/// This can only be used on closures that have the `Copy` trait.
+///
+/// Since this function is implemented for `Fn(...)` closures, we can use
+/// it directly when defining a function:
+/// ```
+/// # use boa_engine::{Context, JsValue, NativeFunction};
+/// # use boa_interop::IntoJsFunctionCopied;
+/// # let mut context = Context::default();
+/// let f = |a: i32, b: i32| a + b;
+/// let f = f.into_js_function_copied(&mut context);
+/// let result = f.call(&JsValue::undefined(), &[JsValue::from(1), JsValue::from(2)], &mut context).unwrap();
+/// assert_eq!(result, JsValue::new(3));
+/// ```
+pub trait IntoJsFunctionCopied<Args, Ret>: private::IntoJsFunctionSealed<Args, Ret> + Copy {
+    /// Converts the type into a JS function.
+    fn into_js_function_copied(self, context: &mut Context) -> NativeFunction;
+}
+
+/// Create a [`JsResult`] from a Rust value. This trait is used to
+/// convert Rust types to JS types, including [`JsResult`] of
+/// Rust values and [`JsValue`]s. It is used to convert the
+/// return value of a function in [`UnsafeIntoJsFunction`] and
+/// [`IntoJsFunctionCopied`].
+pub trait TryIntoJsResult {
+    /// Try to convert a Rust value into a `JsResult<JsValue>`.
+    ///
+    /// # Errors
+    /// Any parsing errors that may occur during the conversion, or any
+    /// error that happened during the call to a function.
+    fn try_into_js_result(self, context: &mut Context) -> JsResult<JsValue>;
+}
+
+mod try_into_js_result_impls;
+
+/// Create a Rust value from a JS argument. This trait is used to
+/// convert arguments from JS to Rust types. It allows support
+/// for optional arguments or rest arguments.
+pub trait TryFromJsArgument: Sized {
```

Comment:
Thought: A bit worried that this allows having two or more `JsRest` arguments on a function, but I'm really not sure what we could do to disallow this at the type level.

Reply:
Yes, I thought about it. Having 2+ `JsRest` basically means the other ones are empty.

What I could also do is the same trick I got with `Context`; only allow `JsRest` at the end of the arguments. WDYT?

Reply:
That would be pretty great! And it makes sense to always require to declare the rest arguments last, so it checks out.

Reply:
Here's another alternative, rename `JsRest` to `JsAll<T> for T: TryFromJs` and the first value to fail the `TryFromJs` will return. WDYT?

Reply:
That way I can do `JsAll<JsString>` to get all the strings in the argument. Or `JsAll<JsValue>` to get the rest of arguments.

Reply:
Sounds fine, but how does that solve the problem of specifying two or more rest arguments?

Reply:
Alright I did both. You can still do `JsAll<i32>, JsAll<i32>` but that's user error I'd say.

Reply:
Sounds good to me!